codalab/model/bundle_model.py

```diff
@@ -633,19 +633,26 @@ def add_join(table: Table, condition: Any, left_outer_join: bool = False):
 
         if user_id != self.root_user_id:
             # Restrict to the bundles that we have access to.
+            aliased_group_bundle_permission = aliased(cl_group_bundle_permission)
             add_join(
-                cl_group_bundle_permission,
-                cl_bundle.c.uuid == cl_group_bundle_permission.c.object_uuid,
+                aliased_group_bundle_permission,
+                cl_bundle.c.uuid == aliased_group_bundle_permission.c.object_uuid,
+                left_outer_join=True,
+            )
+            aliased_cl_user_group = aliased(cl_user_group)
+            add_join(
+                aliased_cl_user_group,
+                aliased_cl_user_group.c.user_id == user_id,
+                left_outer_join=True,
             )
-            add_join(cl_user_group, cl_user_group.c.user_id == user_id, left_outer_join=True)
             access_via_owner = cl_bundle.c.owner_id == user_id
             access_via_group = and_(
                 or_(  # Join constraint (group)
-                    cl_group_bundle_permission.c.group_uuid
+                    aliased_group_bundle_permission.c.group_uuid
```

Comment:
Actually, is there a potential false positive case here (access_via_group is incorrectly true) if `aliased_group_bundle_permission.c.group_uuid` and `self.public_group_uuid` are both `None`?

Reply:
Thinking about this a bit more, this logic seems incorrect?

- `aliased_group_bundle_permission` is all the group bundle permissions for the object
- `aliased_cl_user_group` is all the group memberships for the user
- `aliased_cl_user_group.c.user_id == user_id` doesn't do anything because this is already enforced by the join.
- So it seems like if the bundle grants access to some group A, and the user X belongs to some group B, then user X incorrectly gets access to the bundle?
- Seems like what we're missing here is we need to check that `aliased_group_bundle_permission.c.group_id == aliased_cl_user_group.c.group_id`?
- Alternatively you can push this constraint into the join condition i.e. when you in onto `aliased_group_bundle_permission`, join using the condition that the bundle permission group is the user's groups or the public group only?

Reply:
Yes, you're totally correct. Thanks for catching this! I updated it. I believe it should be correct now.